### PR TITLE
Story/stry0038018

### PR DESCRIPTION
**Pull request by csmcgee**

### CODE REVIEW

File: kong/plugins/oidc/handler.lua [hunk 1]
```diff
@@ -7,7 +7,9 @@ local cjson = require("cjson")
 local openidc = require("resty.openidc")
 
 OidcHandler.PRIORITY = 1000
-
+-- GET /api/groups
```
Comment (csmcgee) [inline kong/plugins/oidc/handler.lua]: @iankoplowitzdev remove comments

File: kong/plugins/oidc/handler.lua [hunk 2]
```diff
@@ -31,6 +33,9 @@ end
 function handle(oidcConfig, oidcSessionConfig)
   local response
 
+  -- clear important request headers to prevent ability to provide them client side
```
Comment (csmcgee) [inline kong/plugins/oidc/handler.lua]: @iankoplowitzdev suggest rewording comment to something like below.
```
-- clear oidc plugin headers to prevent spoofing of info to upstream api
```

File: kong/plugins/oidc/handler.lua [hunk 3]
```diff
@@ -93,10 +98,16 @@ function make_oidc(oidcConfig, oidcSessionConfig)
     end
   end
 
+  -- unauth_action = pass (default)
```
Comment (csmcgee) [inline kong/plugins/oidc/handler.lua]: @iankoplowitzdev remove these comments

Comment (csmcgee) [reply kong/plugins/oidc/handler.lua]: and/or change them to be descriptive

File: kong/plugins/oidc/handler.lua [hunk 4]
```diff
@@ -93,10 +98,16 @@ function make_oidc(oidcConfig, oidcSessionConfig)
...
+  -- unauth_action = nil for /api/auth/bearer/login
+  local unauth_action = "pass"
```
Comment (csmcgee) [inline kong/plugins/oidc/handler.lua]: @iankoplowitzdev `unauth_action` still static - need to make it configurable.

File: kong/plugins/oidc/handler.lua [hunk 5]
```diff
@@ -93,10 +98,16 @@ function make_oidc(oidcConfig, oidcSessionConfig)
...
   -- grab X-Requested-With Header to see if request was from browser/ajax
-  local unauth_action = nil
   local ngx_headers = ngx.req.get_headers()
-  if ngx_headers then
+  -- @TODO: move the hard coded path to config file
+  if kong.request.get_path() == oidcConfig.idp_authentication_path then
```
Comment (csmcgee) [inline kong/plugins/oidc/handler.lua]: @iankoplowitzdev nice use of `kong.request.get_path()` but I think we should use `ngx.var.request_uri` for now to be consistent with the code.

Currently no kong methods are called interesting enough.

Until we learn more about lua and path of plugin would rather stay consistent with current code then add this.  This will also remove the need for the other mock.

File: test/unit/test_handler_mocking_openidc.lua [hunk 6]
```diff
@@ -405,6 +406,45 @@ function TestHandler:test_authenticate_with_session_cookie_samesite_set_to_none(
...
+  end
+  -- act
+  self.handler:access({ idp_authentication_path = idpAuthPath })
+
+  -- assert
+  lu.assertTrue(self:log_contains("login request detected"))
+  lu.assertEquals(actual_unauth_action, nil)
+end
+
+function TestHandler:test_authenticate_ok_with_api_request()
```
Comment (csmcgee) [inline test/unit/test_handler_mocking_openidc.lua]: @iankoplowitzdev  change `_api_request` to `_xmlhttprequest`.

Comment (iankoplowitzdev) [reply test/unit/test_handler_mocking_openidc.lua]: This method will check any api request, not just Ajax calls.

File: README.md [hunk 7]
```diff
@@ -45,6 +52,9 @@ ngx.ctx.authenticated_consumer = {
 }
 ```
 
+### XMLHttp/Ajax Requests
+
+HTTP Requests made by client-side code (e.g ajax) should include the `X-Requested-With: XMLHttpRequest` header. 302 Redirects are replaced with 401 Unauthorized HTTP responses when this header is present AND the user is unauthenticated.
```
Comment (brainthinks) [inline README.md]: why are 302s being turned into 401s?

File: README.md [hunk 8]
```diff
@@ -31,6 +31,13 @@ The diagram below shows the message exchange between the involved parties.
 
 ![alt Kong OIDC flow](docs/kong_oidc_flow.png)
 
+For security purposes the following headers are stripped at the beginning of this plugins execution:
+* `X-Access-Token`
+* `X-ID-Token`
+* `X-Userinfo`
```
Comment (brainthinks) [inline README.md:37]: are these hard coded, or are they part of some configurable list?  shouldn't the be configurable?

File: README.md (hunk 7 above)
Comment (brainthinks) [reply README.md]: also, shouldn't the `force_authentication_path` be taken into consideration before returning a 401?

File: README.md [hunk 9]
```diff
@@ -206,7 +229,7 @@ Server: kong/0.11.0
 
 ### Upstream API request
 
-The plugin adds an additional `X-Userinfo`, `X-Access-Token` and `X-Id-Token` headers to the upstream request, which can be consumer by upstream server. All of them are base64 encoded:
+The plugin adds an additional `X-Userinfo`, `X-Access-Token` and `X-Id-Token` headers to the upstream request, which can be consumer by upstream server. Note if these headers were present in the request prior to the execution fo this plugin, then they will be removed/overwritten. All of them are base64 encoded:
```
Comment (brainthinks) [inline README.md]: of, not fo

File: docs/src/pass_flow.txt [hunk 10]
```diff
@@ -0,0 +1,22 @@
+title Kong OIDC Plugin - "pass" functionality
```
Comment (brainthinks) [inline docs/src/pass_flow.txt]: this file has a .txt extension - is it plantuml?  mermaid?  shouldn't it have a corresponding file extension?  or does github expect .txt?

File: kong/plugins/oidc/handler.lua [hunk 11]
```diff
@@ -93,22 +95,23 @@ function make_oidc(oidcConfig, oidcSessionConfig)
...
-  if ngx_headers then
-    local xhr_value = ngx_headers["X-Requested-With"]
-    -- was the request ajax/async?
-    if xhr_value == "XMLHttpRequest" then
-      -- reference: https://github.com/zmartzone/lua-resty-openidc/blob/master/lib/resty/openidc.lua#L1436
-      -- set to deny so resty.openidc returns instead of redirects (ends request)
-      ngx.log(ngx.DEBUG, "OidcHandler ajax/async request detected, setting unauth_action = deny")
-      unauth_action = "deny"
-    end
+  local unauth_action = oidcConfig.force_authentication_path and "pass" or nil
```
Comment (brainthinks) [inline kong/plugins/oidc/handler.lua]: it seems as though the initialized value of this is being set here.  shouldn't the condition here be alongside the conditions below?  to keep all the logic that sets this value in one place

File: kong/plugins/oidc/utils.lua [hunk 12]
```diff
@@ -132,4 +133,10 @@ function M.cache_get(type, key)
   return value
 end
 
+function M.clear_request_headers()
+  ngx.req.clear_header("X-Access-Token")
+  ngx.req.clear_header("X-ID-Token")
+  ngx.req.clear_header("X-Userinfo")
```
Comment (brainthinks) [inline kong/plugins/oidc/utils.lua]: should these be hard coded?  how likely are they to need to be changed / added to in the future?  can we read them from a configurable list instead?

File: test/unit/test_handler_mocking_openidc.lua [hunk 13]
```diff
@@ -405,6 +406,67 @@ function TestHandler:test_authenticate_with_session_cookie_samesite_set_to_none(
...
+
+  -- assert
+  lu.assertTrue(self:log_contains("login request detected"))
+  lu.assertEquals(actual_unauth_action, nil)
+end
+
+function TestHandler:test_authenticate_ok_to_non_force_authentication_path()
+  -- arrange
+  local actual_unauth_action
+
```
Comment (brainthinks) [inline test/unit/test_handler_mocking_openidc.lua:457]: is there no need to define a request uri here?

File: test/unit/test_handler_mocking_openidc.lua [hunk 14]
```diff
@@ -405,6 +406,67 @@ function TestHandler:test_authenticate_with_session_cookie_samesite_set_to_none(
...
+    return {}, false, "/", session
+  end
+  -- act
+  self.handler:access({ force_authentication_path = idpAuthPath })
+
+  -- assert
+  lu.assertEquals(actual_unauth_action, "pass")
+end
+
+function TestHandler:test_authenticate_ok_to_force_authentication_path()
```
Comment (brainthinks) [inline test/unit/test_handler_mocking_openidc.lua]: this function has the same name - is that ok?

File: README.md (hunk 7 above)
Comment (brainthinks) [reply README.md]: i just read the tests - i definitely think more "what" and "why" would be helpful here

File: test/unit/test_utils.lua [hunk 15]
```diff
@@ -77,4 +78,25 @@ function TestUtils:testDiscoveryOverride()
...
 end
 
+function TestUtils:testClearRequestHeaders()
+  -- assign
+  local headers = {}
+
+  _G.ngx = {
+    req = {
+      clear_header = function(header)
+        headers[header] = true
```
Comment (brainthinks) [inline test/unit/test_utils.lua:89]: if the name of the function indicates clearing a value, wouldn't it be better to use a value like `nil` here, rather than a value that's truthy?  i dunno if it's harder to write an assertion for `nil` in lua

File: kong/plugins/oidc/handler.lua (hunk 11 above)
Comment (brainthinks) [reply kong/plugins/oidc/handler.lua]: seems like this is where the actual magic for this feature occurs.  it may be worth having additional comments.  why are xhr requests denied?  why are non-xhr requests passed?  why is it nil when the force auth path is found?

File: test/unit/test_utils.lua (hunk 15 above)
Comment (csmcgee) [reply test/unit/test_utils.lua:89]: @brainthinks I see your point but we are just making sure `clear_header` is called with 3 different values. That is all this unit test is testing.

Not sure if it is necessary to make it `nil` for a stubbed function.

File: test/unit/test_handler_mocking_openidc.lua (hunk 13 above)
Comment (csmcgee) [reply test/unit/test_handler_mocking_openidc.lua:457]: @brainthinks request uri is checked against `ngx.var.request_uri` which is set in mocak_case.lua:16. `/` by default.

Comment (iankoplowitzdev) [reply test/unit/test_handler_mocking_openidc.lua:457]: It's defined as "/" in the mockable_case.lua file, which is a super type of this function.

File: README.md (hunk 9 above)
Comment (csmcgee) [reply README.md]: addressed

File: docs/src/pass_flow.txt (hunk 10 above)
Comment (csmcgee) [reply docs/src/pass_flow.txt]: addressed, changed extension

File: README.md (hunk 7 above)
Comment (csmcgee) [reply README.md]: @brainthinks added why - hopefully that provides more clarity.

File: README.md (hunk 8 above)
Comment (csmcgee) [reply README.md:37]: @brainthinks these are definitely hardcoded - I don't think we should change them or make them configurable.

Don't think we should consider making them configurable till we got a use case.

File: test/unit/test_handler_mocking_openidc.lua (hunk 14 above)
Comment (csmcgee) [reply test/unit/test_handler_mocking_openidc.lua]: addressed

File: kong/plugins/oidc/handler.lua (hunk 11 above)
Comment (iankoplowitzdev) [reply kong/plugins/oidc/handler.lua]: Updated the logic. Take a look and see if it matches more with what you'd expect.

File: kong/plugins/oidc/utils.lua (hunk 12 above)
Comment (iankoplowitzdev) [reply kong/plugins/oidc/utils.lua]: Updated to use constants. Chris and I agree these aren't values that will likely every be changed/added to.

File: test/unit/test_utils.lua (hunk 15 above)
Comment (brainthinks) [reply test/unit/test_utils.lua:89]: to me it is counter-intuitive to test that a "cleared value" is true or truthy.  I understand that the value is arbitrary given that this is a mock/stub for a test.  up to you if you want to change it.

File: README.md [hunk 16]
```diff
@@ -54,7 +54,11 @@ ngx.ctx.authenticated_consumer = {
 
 ### XMLHttp/Ajax Requests
 
-HTTP Requests made by client-side code (e.g ajax) should include the `X-Requested-With: XMLHttpRequest` header. 302 Redirects are replaced with 401 Unauthorized HTTP responses when this header is present AND the user is unauthenticated.
+XMLHttpRequests made by client-side code (i.e ajax) should include the `X-Requested-With: XMLHttpRequest` header. 302 Redirects are replaced with 401 Unauthorized HTTP responses when this header is present AND the user is unauthenticated.
+
+#### Why?
+
+302 redirects are followed transparently via XMLHttpRequests (xhr/ajax requests) thus there is nothing the client side can do to detect if a 302 happened. Returning a status code of 401 allows the client to respond to the request accordingly.
```
Comment (brainthinks) [inline README.md:61]: In Claris, we include a body with the 401 that contains the redirect location - can / should we do that here, so that the caller has the option of requesting the redirected url?

File: kong/plugins/oidc/handler.lua [hunk 17]
```diff
@@ -96,16 +97,23 @@ function make_oidc(oidcConfig, oidcSessionConfig)
...
+  -- If the request is not ajax, and matches the configured authentication path (redirect user if authentication fails)
   elseif ngx.var.request_uri == oidcConfig.force_authentication_path then
     ngx.log(ngx.DEBUG, "OidcHandler force_authentication_path matched request, setting unauth_action = nil")
-    unauth_action = nil
+    unauth_action = constants.UNAUTH_ACTION.NIL
+
+  -- if force_authentication_path is set then allow requests upstream even if unauthenticated
+  -- if force_authentication_path is NOT set then redirect user if not authenticated
+  else
+    unauth_action = oidcConfig.force_authentication_path and constants.UNAUTH_ACTION.PASS or constants.UNAUTH_ACTION.NIL
```
Comment (brainthinks) [inline kong/plugins/oidc/handler.lua]: this is still a nested condition - it looks like you copied and pasted the exact same line from above into an else condition.

I was thinking something like:

```
elseif oidcConfig.force_authentication_path
  unauth_action = pass
else
  unauth_action = nil
```

Or if nil is set as the initial value, the else block becomes unecessary

this way, each individual value has a path / flow (increases consitency, improves flow, allows for individually-blocked comments)

it is also possible that i'm misunderstanding the `and` syntax in lua

Comment (iankoplowitzdev) [reply kong/plugins/oidc/handler.lua]: Lol... I literally had that solution typed up, but changed it because I thought what I committed is more what you were going for. I'll talk with Chris about it in the am.

File: test/unit/test_utils.lua (hunk 15 above)
Comment (iankoplowitzdev) [reply test/unit/test_utils.lua:89]: It's really not a big issue to change; we can certainly switch it to nil if it makes more sense. But, just for understanding, we're not testing whether the cleared value is truthy, we're just using Lua's hash table implementation to store occurrences of a header, and using the boolean (in this case, "true") to in order to align with what the assertion is expecting. I had a tough time coming up with that last sentence as an explanation. Did it make sense?

Comment (csmcgee) [reply test/unit/test_utils.lua:89]: @brainthinks also "Also, the value of an undeclared variable is nil, because there is no special undefined value. "

see https://www.luafaq.org/gotchas.html

So I think making it nil or undefined or blank adds complexity where it isn't necessary.

Again, we are just making sure the method is called at least 3 times with 3 specific values.
